docker/startup.sh

```diff
@@ -1,4 +1,14 @@
 #!/bin/sh
 # The purpose of this script is to set a log regex to forward error logs to stderr and other logs to
 # stdout. This allows us to easily configure alerts when any message goes to stderr.
-"$@" | regex-stream-split "^\d+-\d+-\d+T\d+:\d+:\d+\.\d+Z\s+(TRACE|DEBUG|INFO|WARN)" "^\d+-\d+-\d+T\d+:\d+:\d+\.\d+Z\s+ERROR"
+
+# Run both commands in the background, so that we can receive signals and forward it to the main process (for this we capture the pid).
+# The stream-split command deliberately ignores interrupts (otherwise we may lose logs on teardown). It terminates automatically
+# once the main command (stdin) terminates.
+( "$@" & echo $! > main_pid ) | (trap '' TERM INT; regex-stream-split '^\d+-\d+-\d+T\d+:\d+:\d+\.\d+Z\s+(TRACE|DEBUG|INFO|WARN)' '^\d+-\d+-\d+T\d+:\d+:\d+\.\d+Z\s+ERROR') &
```

Comment:
@fedgiac Do you mind taking a look as the bash :mage:?

Reply:
I'm by no means an expert of bash traps but I'll try!

I think there is a race condition between `echo $! > main_pid` and `pid=$(cat main_pid)`.
This should be resolved for the most part by using `trap 'kill -15 $(cat main_pid)' TERM`. The traps can then be moved to the start, and the `&` at the end of the current first line and the `wait` can be removed.

After this I don't see anything else that could be troublesome but Bash surprised me many times. :see_no_evil:
As a nit I propose to use `kill -SIGINT`/`-SIGTERM` instead of `-2`/`-15`.

Reply:
> I think there is a race condition between echo $! > main_pid and pid=$(cat main_pid).

Does `echo $! > main_pid` not block until the file is written? Will change to your suggestion anyways.

I don't think the traps can be moved though, since `sh` doesn't seem to be forwarding SIGTERM to the first command of the pipe while it's waiting on `regex-stream-split` to finish. If we spawn `regex-stream-split` as a background task it works however.

My test plan is to run `docker stop` against a running container and expect to see `Gracefully shutting down API`

Reply:
> Does echo $! > main_pid not block until the file is written?

I don't know, but the issue is a bit different: to my understanding the syntax `command >file &` spawns a new thread, and then this new thread writes to `file`. At the same time, the thread that spawned the new thread tries to read the same file immediately afterwards.

As a minimal example of the issue I have this command that prints `before`, `after` or nothing randomly when trying multiple times.
If it doesn't work for you, you can try with other initial values rather than `i=300`.
```
sh -c 'echo before >/tmp/test; ( i=300; while (( i-- )); do :; done && echo after >/tmp/test ) & cat /tmp/test'
```

Reply:
Anyway, how it is right now looks good to me!